### PR TITLE
Add documentation for linker errors around `static` functions and linker scripts

### PR DESCRIPTION
**Pull request by randomPoison**
This (at least partially) addresses #414 and #415 by adding docs around those potential error cases. This covers the cases that I ran into working on zlib.

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -182,6 +182,20 @@ IA2_CALL(opaque, id) // Call the the wrapper that `opaque` points to
...
+## `static` Function Pointers
+
+If a `static` function is used as an address-taken function for cross-compartment calls, then the rewriter needs to add `__attribute__((used))` to the function declaration in order for it to link correctly with the generated callgates. Normally the rewriter can achieve this, but if the `static` keyword comes from a macro the rewriter will not be able to insert the attribute. For example:
+
+```c
+#define local static
+local void function() {}
+```
+
+The same issue can happen when using
```
Comment (randomPoison) [inline docs/usage.md:194]: Woops, I accidentally a

File: docs/usage.md [hunk 2]
```diff
@@ -249,3 +266,38 @@ initializes and makes use of TLS. If you're seeing compartment violations when
...
+may run into linker errors where `libcallgates.so` can't find symbols within a
+compartment. There are a couple of reasons why that may happen.
+
+As mentioned in the rewriter section, if a function is `static`, but the
+`static` keyword comes from inside a macro, then the rewriter will fail to emit
+an `__attribute__((used))` annotation for the function. This can lead to linker
+errors if the function is used as an address-taken function across compartments.
+
+Similar issues can occur when using linker scripts. If your project uses a
+linker script to mark functions local, you'll run into linker errors where the
```
Comment (kkysen) [inline docs/usage.md:282]: What's meant by "local" here exactly? `static`, or does it mean something similar but slightly different?

File: docs/usage.md [hunk 3]
```diff
@@ -249,3 +266,38 @@ initializes and makes use of TLS. If you're seeing compartment violations when
...
+
+If you're seeing errors along the lines of:
+
+```txt
+/usr/bin/ld: libcallgates.so: undefined reference to `foo'
+```
+
+Then IA2 needs `foo` to not be marked local.
+
+Note also that an `__attribute__((visibility ("hidden")))` annotation can have
```
Comment (kkysen) [inline docs/usage.md:301]:
```suggestion
Note also that an `__attribute__((visibility("hidden")))` annotation can have
```

File: docs/usage.md [hunk 4]
```diff
@@ -249,3 +266,38 @@ initializes and makes use of TLS. If you're seeing compartment violations when
...
+the same effect as hiding the symbol in a linker script. Such annotation will
+have to be removed to work in a containerized app.
```
Comment (kkysen) [inline docs/usage.md:303]: We figured out how to handle hidden symbols in #451.  Though it appears #435 was then re-opened, but I forget why (@ayrtonm, do you remember?).

Comment (fw-immunant) [inline docs/usage.md:303]: "compartmentalized", not "containerized"

File: docs/usage.md (hunk 2 above)
Comment (fw-immunant) [reply docs/usage.md:282]: I don't know the canonical definition of the notion of a local symbol, but this link is a decent starting point: https://sourceware.org/binutils/docs/ld.html#VERSION

At the very least, one difference is that `static` is a C notion, while local symbols are a linker notion.

Comment (fw-immunant) [reply docs/usage.md:282]: It might be clearer to say "mark function **symbols** as local".

File: docs/usage.md (hunk 4 above)
Comment (ayrtonm) [reply docs/usage.md:303]: It was closed by mistake. There's some edge case we still don't support but I don't remember at the moment. IIRC adding the flag to the test suite should point it out.

Comment (randomPoison) [reply docs/usage.md]: Hrm, interesting I didn't realize we had already resolved part of this. I'm trying to reproduce the issues I'm documenting here and it seems like maybe these docs are now redundant?

* `__attribute__((used))` is now being inserted correctly for static functions when `static` comes from inside a macro.
* Functions marked with `__attribute__((visibility ("hidden")))` now link correctly.

I still need to test if the issue with a version script is also fixed, but if it is I think we can close this PR.

Comment (randomPoison) [reply docs/usage.md]: Looks like the fix for `visibility("hidden")` also addressed the issue with having a linker script mark a symbol `local`? Or at least the issue was fixed at some point because I can now include the zlib linker script in the compartmentalized build and I no longer hit the linker errors I originally saw. So I'm going to close this and the related bugs.

**Comment by randomPoison** [comment]
The issues this PR documents no longer seem to be happening (see https://github.com/immunant/IA2-Phase2/pull/484#discussion_r1932993068), so I'm going to close this PR.
